proposals/ignored-directives.md

```diff
@@ -0,0 +1,126 @@
+# Ignored directives
+
+Champion issue: <https://github.com/dotnet/csharplang/issues/3507>
+
+## Summary
+
+Add `#` directives to be used by tooling, but ignored by the language.
+
+```cs
+#!/usr/bin/dotnet run
+#sdk Microsoft.NET.Sdk.Web
+#package Microsoft.AspNetCore.OpenApi 9.0.2
+
+var builder = WebApplication.CreateBuilder(args);
+
+builder.Services.AddOpenApi();
+
+var app = builder.Builder();
+
+app.MapGet("/", () => new { Message = "Hello, World!" })
+    .WithName("HelloWorld");
+
+app.Run();
+```
+
+## Motivation
+
+We are adding `dotnet run file.cs` support [in .NET SDK][dotnet-run-file].
+Real-world file-based programs need to reference NuGet packages.
+It would be also useful if it were possible to execute file-based programs directly like `./file.cs` when they have [the shebang directive][shebang] (`#!`).
+
+The language should ignore these directives, but compiler implementations and other tooling can recognize them.
+We already have similar directives in the language:
+- `#region`
+- `#pragma`: details are implementation specific
+- `#error version`: the `version` part is not in the spec, but Roslyn will report its version
+
+## Detailed design
+
+Introduce new *ignored pre-processing directives* ([§6.5][directives]):
+
+```antlr
+PP_Kind
+    : ... // Existing directive kinds
+    | PP_Ignored
+    ;
+
+PP_Ignored
+    : PP_IgnoredToken Input_Character*
+    ;
+
+PP_IgnoredToken
+    : '!'
+    | 'package'
+    | 'sdk'
+    ;
+```
+
+### Restrictions
+
+Ignored directives must occur before the first token ([§6.4][tokens]) in the compilation unit, just like `#define`/`#undef` directives.
+This improves readability (all package references and other configuration is in one place), tooling performance (no need to scan long files in full),
+and simplifies the design (no need to figure out what would be the meaning of ignored directives inside `#if false`).
+
+Furthermore, the compiler should report a warning if the `#!` directive is not placed at the first line and the first character in the file
+(not even a BOM marker can be in front of it), because otherwise shells won't recognize it.
```

Comment:
not a complaint.  but we will have to likely update our tools that deal with headers (and determining if a doc is generated) to let them know that this info must be first, and nothing can precede it.